### PR TITLE
[Performance](Variant) Improve load performance for variant type

### PR DESCRIPTION
**Pull request by eldenmoon**
1. remove phmap for padding rows
2. add `SimpleFieldVisitorToScarlarType` for short circuit type deducing
3. correct type coercion for conflict types bettween integers
4. improve nullable column performance
5. remove shared_ptr dependancy for DataType use TypeIndex instead
6.  Optimization by caching the order of fields (which is almost always the same)
    and a quick check to match the next expected field, instead of searching the hash table.

In clickbench data, load performance:
12m36.799s ->7m10.934s about 43% latency reduce

In variant_p2/performance.groovy:
3min44s20 -> 1min15s80 about 66% latency reducy
 
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Review by airborne12 — APPROVED** [review]
LGTM

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run p0

File: be/src/vec/common/schema_util.cpp [hunk 1]
```diff
@@ -114,31 +114,8 @@ bool is_conversion_required_between_integers(const IDataType& lhs, const IDataTy
...
-    }
-    if (lhs == FieldType::OLAP_FIELD_TYPE_SMALLINT) {
-        return !(rhs == FieldType::OLAP_FIELD_TYPE_TINYINT ||
-                 rhs == FieldType::OLAP_FIELD_TYPE_SMALLINT);
-    }
-    if (lhs == FieldType::OLAP_FIELD_TYPE_TINYINT) {
-        return !(rhs == FieldType::OLAP_FIELD_TYPE_TINYINT);
-    }
-    return true;
+    return (!is_native_int && !is_native_uint) ||
```
Comment (xiaokang) [inline be/src/vec/common/schema_util.cpp:117]: opposite to original code?

File: be/src/vec/columns/column_object.cpp [hunk 2]
```diff
@@ -148,6 +151,84 @@ class FieldVisitorToNumberOfDimensions : public StaticVisitor<size_t> {
     }
 };
 
+// Visitor that allows to get type of scalar field
+// but include contain complex field.This is a faster version
+// for FieldVisitorToScalarType which does not support complex field.
+class SimpleFieldVisitorToScarlarType : public StaticVisitor<size_t> {
```
Comment (xiaokang) [inline be/src/vec/columns/column_object.cpp:157]: typo SimpleFieldVisitorToScalarType

File: be/src/vec/columns/column_object.cpp [hunk 3]
```diff
@@ -308,18 +399,29 @@ void ColumnObject::Subcolumn::insert(Field field, FieldInfo info) {
...
-    } else if (!least_common_base_type->equals(*base_type) && !is_nothing(base_type)) {
-        if (!schema_util::is_conversion_required_between_integers(*base_type,
-                                                                  *least_common_base_type)) {
+    } else if (!nested_common_base_type->equals(*nested_base_type) &&
+               !is_nothing(nested_base_type)) {
+        if (schema_util::is_conversion_required_between_integers(*nested_base_type,
+                                                                 *nested_common_base_type)) {
+            LOG_EVERY_N(INFO, 100) << "Conversion between " << base_type->get_name() << " and "
+                                   << least_common_base_type->get_name();
             get_least_supertype<LeastSupertypeOnError::Jsonb>(
```
Comment (xiaokang) [inline be/src/vec/columns/column_object.cpp:423]: Is least_common_base_type changed after calling get_least_supertype? If no, where is get_least_supertype updated?

File: be/src/vec/columns/column_object.cpp (hunk 2 above)
Comment (eldenmoon) [reply be/src/vec/columns/column_object.cpp:157]: fixed

File: be/src/vec/columns/column_object.cpp (hunk 3 above)
Comment (eldenmoon) [reply be/src/vec/columns/column_object.cpp:423]: in `add_new_column_part`

**Comment by eldenmoon** [comment]
run buildall

File: be/src/vec/common/schema_util.cpp (hunk 1 above)
Comment (eldenmoon) [reply be/src/vec/common/schema_util.cpp:145]: yes， the original code logical is wrong

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: be/src/vec/columns/column_object.cpp [hunk 4]
```diff
@@ -220,20 +281,30 @@ class FieldVisitorToScalarType : public StaticVisitor<size_t> {
...
             to_scalar_type_visitor.need_convert_field(),
             apply_visitor(FieldVisitorToNumberOfDimensions(), field),
     };
 }
 
+void get_field_info(const Field& field, FieldInfo* info) {
+    if (field.is_complex_field()) {
+        get_field_info_impl<FieldVisitorToScalarType>(field, info);
+    } else {
+        get_field_info_impl<SimpleFieldVisitorToScalarType>(field, info);
```
Comment (xiaokang) [inline be/src/vec/columns/column_object.cpp:304]: Is there any opt for SimpleFieldVisitorToScalarType?

**Review by xiaokang — APPROVED** [review]

**Pull request merged**